components/layout/fragment.rs

```diff
@@ -2158,17 +2158,37 @@ impl Fragment {
             let block_flow = flow.as_block();
             let start_margin = block_flow.fragment.margin.block_start;
             let end_margin = block_flow.fragment.margin.block_end;
-            if style.get_box().overflow_y == overflow_x::T::visible {
-                if let Some(baseline_offset) = flow.baseline_offset_of_last_line_box_in_flow() {
-                    let ascent = baseline_offset + start_margin;
-                    let space_below_baseline = block_flow.fragment.border_box.size.block -
-                        baseline_offset + end_margin;
-                    return InlineMetrics::new(ascent, space_below_baseline, baseline_offset)
-                }
-            }
-            let ascent = block_flow.fragment.border_box.size.block + end_margin;
-            let space_above_baseline = start_margin + ascent;
-            InlineMetrics::new(space_above_baseline, Au(0), ascent)
+            let border_box_block_size = block_flow.fragment.border_box.size.block;
+
+            //     --------
+            //      margin
+            // top -------- + +
+            //              | |
+            //              | |
+            //  A  ..pogo.. | + baseline_offset_of_last_line_box_in_flow()
+            //              |
+            //     -------- + border_box_block_size
+            //      margin
+            //  B  --------
+            //
+            // § 10.8.1 says that the baseline (and thus ascent, which is the
+            // distance from the baseline to the top) should be A if it has an
+            // in-flow line box and if overflow: visible, and B otherwise.
+
+            let (ascent, space_below_baseline) =
+                match (flow.baseline_offset_of_last_line_box_in_flow(),
+                       style.get_box().overflow_y) {
+                // Case A
+                (Some(baseline_offset), overflow_x::T::visible) => {
+                    (baseline_offset, border_box_block_size - baseline_offset + end_margin)
+                },
+                // Case B
+                _ => (border_box_block_size + end_margin, Au(0))
```

Comment:
Seems like you could derive `space_below_baseline` from `ascent`, using:

```rust
let space_below_baseline = border_box_block_size + end_margin - ascent;
```

I think it would be slightly more readable, what do you think?